src/time_stream/aggregation.py

```diff
@@ -1,573 +1,354 @@
-            unnest_list.extend(aggregation_stage.unnest())
-            with_column_list.extend(aggregation_stage.with_columns())
-
-        group_by_dynamic = df.group_by_dynamic(
-            index_column=self.ts.time_name,
-            every=self.aggregation_period.pl_interval,
-            offset=self.aggregation_period.pl_offset,
-            closed="left",
-        )
-
-        df = group_by_dynamic.agg(aggregation_expression_list)
-
-        if unnest_list:
-            df = df.unnest(*unnest_list)
-        if with_column_list:
-            for with_column in with_column_list:
-                df = df.with_columns(with_column)
-
-        return df
-
-
-class AggregationStage(ABC):
-    """An abstract class used to assist with aggregating data
-
-    Each subclass will add various things to various stages of the aggregation process
-
-    Attributes:
-        aggregator: The PolarsAggregator being aggregated
-        name: The name of this aggregator
-    """
-
-    def __init__(self, aggregator: PolarsAggregator, name: str) -> None:
-        self.aggregator = aggregator
-        self.name = name
+        # If it's already an instance, return it
+        if isinstance(aggregation, AggregationFunction):
+            return aggregation
+
+        # If it's a string, look it up in the registry
+        elif isinstance(aggregation, str):
+            try:
+                return _AGGREGATION_REGISTRY[aggregation]()
+            except KeyError:
+                raise KeyError(f"Unknown aggregation '{aggregation}'.")
+
+        # If it's a class, instantiate it
+        elif isinstance(aggregation, type):
+            if issubclass(aggregation, AggregationFunction):
+                return aggregation()
+            else:
+                raise TypeError(f"Aggregation class {aggregation.__name__} must inherit from AggregationFunction")
 
-    def aggregation_expressions(self) -> list[pl.Expr]:
-        """Return a list of expressions to be applied during the aggregation stage.
-
-        These expressions are passed to the Polars .agg() method.
-
-        Subclasses override this method to provide a non-empty return list.
-
-        Returns:
-            A list of Polars expressions
-        """
-        return []
-
-    def unnest(self) -> list[str]:
-        """Return a list of structs to be unnested after the aggregation stage.
-
-        These struct names are passed to the Polars DataFrame.unnest() method.
+        else:
+            raise TypeError(
+                f"Aggregation must be a string, AggregationFunction class, or instance. "
+                f"Got {type(aggregation).__name__}"
+            )
 
-        Subclasses override this method to provide a non-empty return list.
+    def apply(
+        self,
+        ts: TimeSeries,
+        aggregation_period: Period,
+        columns: Union[str, List[str]],
+        missing_criteria: Optional[Tuple[str, Union[float, int]]] = None,
```

Comment:
`missing_criteria` is a different type than that given in `base`. Need to align.

I think there would be downstream changes in this module if you change here

Reply:
We would need to think about the API aswell because the `missing_criteria` is currently being setup as a `dict` and passed to timestream

https://github.com/NERC-CEH/dri-data-api/blob/385ca2f98e2212f54ce279823a68a45bf0f8086a/src/dataapi/v1/utils.py#L109